diff -u -r orig/src/rcs.c new/src/rcs.c
--- orig/src/rcs.c	2019-04-07 00:51:26.766870234 +0900
+++ new/src/rcs.c	2019-04-07 00:54:35.315881066 +0900
@@ -133,7 +133,7 @@
    string comparisons.  This macro speeds things up a bit by skipping
    the function call when the first characters are different.  It
    evaluates its arguments multiple times.  */
-#define STREQ(a, b) (*(char *)(a) == *(char *)(b) && strcmp ((a), (b)) == 0)
+#define STREQ(a, b) ((a) != NULL && (b) != NULL && *(char *)(a) == *(char *)(b) && strcmp ((a), (b)) == 0)
 
 static char * getfullCVSname (char *, char **);
 
